### DISCUSSION TITLE
Use the AdminJS library for the user interface

### DISCUSSION THREAD

**Discussion by fkrauthan** [keystonejs/keystone#7471] [Feature requests]
I was wondering the other day if it wouldn't be beneficial for Keystone to migrate the admin UI to use [AdminJS](https://docs.adminjs.co/index.html) as this would give a lot more flexibility of configuring the UI. I know you can already manually do that with disabling the native UI and use the `extendExpressApp`. Then just pointing AdminJS to use the Prisma client (and the auto generated Prisma schema). But this obviously will not execute any of the Keystone hooks and permission checks and would have to be replicated. But pre-configuring all of this based on the keystone hooks and give the ability to inject any other AdminJS config could make this extremely powerful.

Curious what other people think.

**Comment by jlarmstrongiv** [comment]
As a user, my thoughts are that switching to https://adminjs.co/ or https://www.prisma.io/studio may free up a lot of attention, but we’d be worse off.

When I think of the native admin ui, it reminds me of https://www.sanity.io/studio a beautiful data manager that includes permission controls and custom blocks.  Quite frankly, AdminJS and other alternatives I know of don’t even come close.  Having a beautiful and functional data manager in open-source frameworks is a rare thing, and I’d be sad to lose it.

**Reply by fkrauthan** [reply]
I don't fully agree. I find AdminJS out of the box has a very similar feeling to what keystone currently provides. But it gives me a lot more flexibility in tailoring the UI even further towards my use-case.

One of the simplest examples is how easy it is to change the form layout (e.g. having two fields next to each other instead of below). Or grouping fields with section headers.

I don't understand what you mean by permission controls as AdminJS seem to have that as well?

**Reply by raveling** [reply]
@fkrauthan 

> how easy it is to change the form layout (e.g. having two fields next to each other instead of below). Or grouping fields with section headers.

It's interesting and valuable for us to understand the kinds of things you want to do with an Admin interface. Feel free to extrapolate on your most desired customisation features, and we can fold that into our research.

Do you have an example of a Keystone powered DB working with Admin JS? If so we'd love to take a peek! 📹

**Comment by raveling** [comment]
@fkrauthan thanks for sharing your thinking here 🙌 

AdminJS looks like a really great project, but we’re committed to keeping Admin UI a core part of the Keystone offering.

> this would give a lot more flexibility of configuring the UI

☝️  this is definitely something we want to solve for developers. Our goal is to offer the same great DX for customising Admin UI that Keystone already has for declaring schema ✨  🙂 

We've been working on the redesign of Admin UI (and related customisation APIs) for a while now, and we're optimistic that the wait will be worth it.

Hope this helps!

**Comment by fkrauthan** [comment]
I think a couple of key useful features are (at least from my point of view). This list contains all my ideas some of them might be already possible.

 
**General**

- Don't display certain entities at all in the admin UI
- Customize the Dashboard with custom react component (so I can show actual stats etc)
- Support for Date only field


**Edit/Show:**

- Having fields in a row (e.g. start date and end date)
- Being able to group fields (e.g. one section for personal information and one section for access information for a user entity)
- Being able to have tabs (e.g. if I have a Book entity I want on first tab meta information (e.g. Name, Author etc) and on a second tab manage the content (especially if repeatable json objects are gonna be implemented (or through references)
- Being able to modify referenced entities from within one entities UI (e.g. Book has Chapters and Chapters has a content field. I want to edit all chapters from within the Book form)


**List**

- Configure which columns to show
- Configure potentially some virtual column where I can define how it will be displayed (e.g. a date in a specific format or instead of list of items (oneToMany) show number of items


I hope this makes sense.

**Reply by raveling** [reply]
Thanks for sharing @fkrauthan we've added this as signal for the design process 🙇

**Reply by raveling** [reply]
@fkrauthan 

> Customize the Dashboard with custom react component (so I can show actual stats etc)

I believe you could achieve this with Custom Pages and Custom Navigation:

https://keystonejs.com/docs/guides/custom-admin-ui-navigation
https://keystonejs.com/docs/guides/custom-admin-ui-pages

Where you replace the dashboard page and nav link with your own thing.

**Reply by raveling** [reply]
> Configure which columns to show

There is a UI to configure columns in the List View:

![image](https://user-images.githubusercontent.com/6447754/166848159-8f8de52d-c6b3-436b-adee-8d8fd9fa7395.png)

**Reply by fkrauthan** [reply]
But is that column config global or just for the current user/browser session? Like for example I would want to be able to not even offer Description as a column (just as an example).

**Reply by raveling** [reply]
You can set things like initial columns, and sort order with Keystone's current capabilities. See `listView` in this section of the docs:

https://keystonejs.com/docs/apis/schema#ui

**Reply by fkrauthan** [reply]
Oh nice :) yeah like I said in my initial comment that some of this things might be already possible. Just included all items for completeness. Unfortunately the missing date field and dynamic drop zone (dynamic object selection) did prevent me from using keystone for my current project but I will definitely keep an eye on this project (as most of this features are at least on the roadmap)

**Reply by raveling** [reply]
> for example I would want to be able to not even offer Description as a column

```ts
export const lists = {
  Post: list({
    fields: {
      title: text({}),
      status: select({
        type: 'enum',
        options: [
          { label: 'Draft', value: 'draft' },
          { label: 'Published', value: 'published' },
        ],
        ui: {
          listView: {
            // hides the field from the listview
            fieldMode: ({ session, context }) => 'hidden',
          },
        },
      }),
    },
    ui: {
      listView: {
        defaultFieldMode: ({ session, context }) => 'read',
        // sets which fields appear in the columns
        initialColumns: ['title', 'content'],
      },
    },
  }),
};
```
